torchao/prototype/quantized_training/int8_mixed_precision.py

```diff
@@ -0,0 +1,247 @@
+    return _Int8MixedPrecisionLinear.apply(act, weight._data, bias, weight.config)
+
+
+@implements(
+    [
+        aten.detach.default,
+        aten.clone.default,
+        aten._to_copy.default,
+        aten.empty_like.default,
+        # FSDP ops
+        aten.slice.Tensor,
+        aten.new_zeros.default,
+        aten.view.default,
+        aten.as_strided.default,
+        c10d_functional.all_gather_into_tensor.default,
+        _c10d_functional.all_gather_into_tensor.default,
+        c10d_functional.wait_tensor.default,
+        _c10d_functional.wait_tensor.default,
+    ]
+)
+def _(func, types, args, kwargs):
+    out = Int8MixedPrecisionLinearWeight(func(args[0]._data, *args[1:], **kwargs), args[0].config)
+    return return_and_correct_aliasing(func, args, kwargs, out)
+
+
+@implements(
+    [
+        aten.copy_.default,
+        aten.addcdiv_.default,
+        aten.add_.Tensor,
+        aten.mul_.Tensor,
+        # param init functions
+        aten.uniform_.default,
+        aten.erfinv_.default,
+        aten.clamp_.default,
+    ]
+)
+def _(func, types, args, kwargs):
+    unpacked_args = [x._data if isinstance(x, Int8MixedPrecisionLinearWeight) else x for x in args]
+    func(*unpacked_args, **kwargs)
+    return args[0]
+
+
+@implements([aten._fused_adam_.default, aten._fused_adamw_.default])
+def _(func, types, args, kwargs):
+    params = [x._data if isinstance(x, Int8MixedPrecisionLinearWeight) else x for x in args[0]]
+    func(params, *args[1:], **kwargs)
+
+
+# return normal tensor
+@implements(
+    [
+        aten.zeros_like.default,  # called by optimizers
+        aten.add.Tensor,
+        aten.mul.Tensor,
+    ]
+)
+def _(func, types, args, kwargs):
+    unpacked_args = [x._data if isinstance(x, Int8MixedPrecisionLinearWeight) else x for x in args]
+    return func(*unpacked_args, **kwargs)
+
+
+# FSDP op
+@implements(aten.split.Tensor)
+def _(func, types, args, kwargs):
+    data_list = func(args[0]._data, *args[1:], **kwargs)
+    return [Int8MixedPrecisionLinearWeight(x, args[0].config) for x in data_list]
+
+
+# alternative UX
+class Int8MixedPrecisionLinear(nn.Linear):
+    def __init__(self, *args, config: Int8MixedPrecisionConfig, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.config = config
+
+    def forward(self, input: Tensor) -> Tensor:
+        return _Int8MixedPrecisionLinear.apply(input, self.weight, self.bias, self.config)
+
+    def extra_repr(self):
+        return f"{super().extra_repr()}, config={self.config}"
+
+    @classmethod
+    def convert_linear(cls, module: nn.Module, config: Int8MixedPrecisionConfig = _DEFAULT_CONFIG):
+        if module.__class__ is nn.Linear:  # exact match, don't swap nn.Linear subclasses
+            module.__class__ = cls
+            module.config = config
+            return
+        for child in module.children():
+            cls.convert_linear(child, config)
+
+
+def _dynamic_int8_linear(input: Tensor, weight: Tensor) -> Tensor:
```

Comment:
I think this should be called `mm` and not `linear`

Reply:
From my microbenchmarks, INT8 mm kernel is most performant with `A @ B.T` i.e. linear, thus I want to only call INT8 kernel with this layout. Need to enforce `A` and `B` in `A @ B.T` to be contiguous though, especially in backward pass. Will double check on this layout+perf stuff.

Also, by only doing `A @ B.T`, I only need to support row-wise scaling for both `A` and `B`.